site/guides/variant-constructor-pattern.md

```diff
@@ -0,0 +1,90 @@
+If you look at the config workflows, you will notice that creating variant of a package is a very common and frequent operation, so reducing the manual steps required in creating a variant can have significant benefits for the package consumers. In this guide, we will look at some techniques/patterns that a package author can use to enable automatic variant construction of a package.
+
+kpt packages comes in two flavors:  `abstract package` and `deployable package instance`. `abstract` package is a reususable package that is used to create deployable package instances that are deployed to a kubernetes cluster. In programming language terms, you can think of `abstract` packages as the classes and  `deployable` packages as the instances of the classes. `deployable` instances of package are also referred as `variant` of the package. 
+
+Figure below shows a `package catalog` on the left that has `abstract` packages and `deployable instances` on the right. A good pattern is to keep the abstract packages and instance packages in separate repos and typically `deployable instances` repo will be setup to auto deploy to a kubernetes cluster using tools such as `config-sync`, `flux`, `argocd`.
+
+![variant constructor pkg repo diagram](img/variant-constructor-pkg-repo-diagram.png)
+
+
+Resources in an `abstract` package have placeholder values that needs to be substituated with actual values to make them ready for deployment. 
+For example, name of the namespace resource below has `example` as a placeholder value. This is a part of the `abstract package`.
+
+```yaml
+apiVersion:v1
+kind: Namespace
+metadata:
+	name: example # <-- this is a placeholder value
+```
+
+
+A kpt package contains kubernetes resources. Whenever you are creating a variant of the package, first step is to ensure unique identity of the resources in that variant. For example, if the abstract package contained a `namespace` resource, then the variant package will have namespace resource corresponding to that variant.
+
+In a kubernetes cluster, resources are identified by their group, version, kind, namespace and name (aka GVKNN). If resource is cluster scoped, then the `metadata.name` uniquely identifies the resource in a cluster. If the resource is namespace scoped, then (namespace, name) together identifies the resource uniquely.
+
+If you look at the steps involved in variant construction, it involves:
+1. set of functions that can help in updating unique identity of the resources
+2. Input to the functions for step 1
```

Comment:
Input generation is one aspect that's critical to the variant-constructor pattern that makes it distinct from other value transformers, so I'd emphasize that more.

Reply:
You are right. Added some more details.